### ISSUE TITLE
Next-gen security - let's make access control smart with behaviour analysis and statistical learning

### ISSUE THREAD

**Issue by aeneasr** [ory/hydra#795]
This issue introduces you to a new set of planned features. There is an immense hype around machine and statistical learning, and we are well aware of its limitations and opportunities. Buffing security with behavioral analysis and machine learning is a tremendous opportunity and a natural step for defenders and blue teams.

The core idea is to equip the exchange and issuance of credentials (such as access and refresh tokens) as well as access control decisions (such as the Warden API and ORY Oathkeeper) with the means to detect malicious actors intelligently and proactively - as opposed to after the data-leak. We have been running a test-system in the shadows for a while now and feel confident enough to introduce this feature to the community.

The system uses several features (request metadata, access request data, ...) to compute a score ranging from 0 (no risk) to 1 (high risk) which can be used to block suspicious requests. It will be consumable through an API which is provided by our infrastructure. The plan is to integrate this API in both ORY Hydra and ORY Oathkeeper and enable it by default. Disabling this feature will be possible.

The system will protect you (reject requests, raise alerts) if it suspects that an access token, refresh token, or user account is compromised. While everything is API-only for now, we plan to introduce management and alert capabilities to this system in the future.

### FAQ

Here are some questions & answers we have encountered in preliminary testing.

#### Will this feature be free?

Yes and no. The feature will be enabled by providing an API key. All our products will have default API keys. These keys are issued under a fair-use policy (to be defined). Deployments exceeding the fair-use policy can either disable this feature or buy a dedicated SLA.

#### Can I get my own API keys?

Yes, for a fee (to be defined).

#### What benefits do I get from dedicated API keys?

We will be able to train our models specifically in the context of your environment. If you are, for example, a bank you will have other behavioral patterns than a freemium app. With a dedicated API key, we will be able to better train our models to the behavior of your users.

#### What data will you collect?

This is still in flux and depends primarily on the fitness of our models. Some things are definitely going to be collected:

* Datetime
* IP
* Location
* Session-ID
* User-ID
* Interaction type (credential issuance, access request, ...)
* Access Control metadata (resource, subject, action, context, ...)
* Request metadata (method, headers)

#### Will you anonymize the data?

We have to find a good balance between making good decisions and anonymizing data. The only personally identifiable information we currently collect is the IP address. As most services (Google Analytics) do this we don't believe that this will be a huge issue for most.

If you wish to have some special anonymization in place, we can probably figure something out as well.

#### Can I define features for your models?

If you have an SLA with us, you will be able to define and use your own features (e.g. user roles, departments, ...).

#### Will this technology become open source?

It is unlikely that this technology will become open source.

#### Can I run this technology on my own infrastructure?

This is currently not planned, but we are open to discuss it as part of a dedicated SLA.

### Now it's your turn

I'm sure you have some questions, feedback, and ideas. Please share them here. Let's get this technology to the next level, and make your security infrastructure as tough as humanly (AI) possibly ;)

**Comment by pnicolcev-tulipretail** [comment]
This sounds like a great way for you to monetize all the hard work you've been doing on Hydra and it's a feature we will seriously consider.

**Comment by MOZGIII** [comment]
Sounds cool.

I'd like this feature to be a kind of extension, that one would be able to replace with his custom implementation if needed. I mean, if one already have a far superior threat detection/analysis system (i.e. integrated with network perimeter/firewalls, in-house IDSes and stuff like that), implementing a custom solution would be the only option (based on the existing investments in the infrastructure).

Now when I think about it - it sounds like a niche product. People that don't care about the data leaks won't care about it (and some will even explicitly disable the feature), those who care might invest heavily in security and already have something more generic deployed (or have it in plans). That leaves us with people who care but don't have existing solution (and they also have to use hydra). A market research is needed to tell for sure, but I don't think there'd a lot of those people.

I'd recommend looking at providing enterprise support and hosted solution for hydra (basically like Auth0, but with core product being open source). Lately Gitlab had huge success with model like this. I'm not an very familiar with all this though.

In the technical perspective, what I really don't like is the API keys and stuff being baked in (including default keys). That means for some use cases one would have to fork the code and remove the API keys and remote server interaction parts (just like some people can find it necessary to clear out telemetry).

Also, note hydra, being sort of a generic OAuth token manager, may be used to not authenticate people, meaning `sub` not necessarily being related to a person. For example, someone might use it to authenticate a device (i.e. a teapot, or a drone, or a tank).

To summarize, I like the idea, and the fact that it will allow to monetize hydra and to help you keep up the good work, but also I'm kind of afraid that those additions, if executed in forced and hard-to-cut-out way, might make hydra less appealing to use for me. I have use cases where I'd use the hosted AI threat detection feature, but I also have some where I'd need to cut it out. I also have cases where I don't care about the security at all. 🙄 

Hope my feedback is useful.

**Comment by arekkas** [comment]
> I'd like this feature to be a kind of extension, that one would be able to replace with his custom implementation if needed. I mean, if one already have a far superior threat detection/analysis system (i.e. integrated with network perimeter/firewalls, in-house IDSes and stuff like that), implementing a custom solution would be the only option (based on the existing investments in the infrastructure).

Maybe, not sure - I think one could solve it quite easily with a reverse proxy. Usually these systems should have something like that in place already! But it's definitely optional!

> Now when I think about it - it sounds like a niche product. People that don't care about the data leaks won't care about it (and some will even explicitly disable the feature), those who care might invest heavily in security and already have something more generic deployed (or have it in plans). That leaves us with people who care but don't have existing solution (and they also have to use hydra). A market research is needed to tell for sure, but I don't think there'd a lot of those people.

Yeah, that's definitely something worth thinking of. We have not a good idea how large the user base would be, and have only few data points that aren't really meaningful imo.

> I'd recommend looking at providing enterprise support and hosted solution for hydra (basically like Auth0, but with core product being open source). Lately Gitlab had huge success with model like this. I'm not an very familiar with all this though.

Two pieces:

* Support: Doesn't really scale because as you grow your margins shrink (as more management is required), unless you increase the price. Sure, there are companies like RedHat and so on, but they are really the exception.
* Hosting: Only start ups want this hosted, almost no enterprise wants to give control over their access management away.

> To summarize, I like the idea, and the fact that it will allow to monetize hydra and to help you keep up the good work, but also I'm kind of afraid that those additions, if executed in forced and hard-to-cut-out way, might make hydra less appealing to use for me. I have use cases where I'd use the hosted AI threat detection feature, but I also have some where I'd need to cut it out. I also have cases where I don't care about the security at all. 🙄

I agree, and the points you raise all make sense. The biggest hurdle to implementing this right now for us is to:

1. Understand demand and pricing
2. Get Hydra 1.0.x out

Thank you for your input! It's really appreciated!
